Modules/posixmodule.c

```diff
@@ -3330,85 +3332,166 @@ static PyMemberDef pystatx_result_members[] = {
-        "maximum optimized size for direct I/O with torn-write protection"),
+STATX_GET_UINT(stx_atomic_write_unit_max_opt, stx_atomic_write_unit_max_opt,
+               STATX_WRITE_ATOMIC)
 #endif
-#ifdef STATX_DIO_READ_ALIGN
-    MM(stx_dio_read_offset_align, Py_T_UINT, dio_read_offset_align,
-        "direct I/O file offset alignment for reads"),
+
+
+#define STATX_GET_ULONGLONG(ATTR, MEMBER, MASK) \
+    static PyObject* \
+    pystatx_result_get_##ATTR(PyObject *op, void *Py_UNUSED(context)) \
+    { \
+        Py_statx_result *self = Py_statx_result_CAST(op); \
+        if (!(self->stx.stx_mask & MASK)) { \
+            Py_RETURN_NONE; \
+        } \
+        unsigned long long value = self->stx.MEMBER; \
+        return PyLong_FromUnsignedLongLong(value); \
+    }
+
+STATX_GET_ULONGLONG(st_blocks, stx_blocks, STATX_BLOCKS)
+STATX_GET_ULONGLONG(st_ino, stx_ino, STATX_INO)
+STATX_GET_ULONGLONG(st_size, stx_size, STATX_SIZE)
+#ifdef STATX_MNT_ID
+STATX_GET_ULONGLONG(stx_mnt_id, stx_mnt_id, STATX_MNT_ID)
+#endif
+#ifdef STATX_SUBVOL
+STATX_GET_ULONGLONG(stx_subvol, stx_subvol, STATX_SUBVOL)
 #endif
-    {NULL},
-};
 
-#undef MX
-#undef MM
-#undef M
 
-static PyObject *
-pystatx_result_get_nsec(PyObject *op, void *context)
-{
-    uint16_t offset = (uintptr_t)context;
-    struct statx_timestamp *ts = (void*)op + offset;
-    _posixstate *state = PyType_GetModuleState(Py_TYPE(op));
-    assert(state != NULL);
-    return stat_nanosecond_timestamp(state, ts->tv_sec, ts->tv_nsec);
-}
+#define STATX_GET_DOUBLE(ATTR, MEMBER, MASK) \
+    static PyObject* \
+    pystatx_result_get_##ATTR(PyObject *op, void *Py_UNUSED(context)) \
+    { \
+        Py_statx_result *self = Py_statx_result_CAST(op); \
+        if (!(self->stx.stx_mask & MASK)) { \
+            Py_RETURN_NONE; \
+        } \
+        double sec = self->MEMBER; \
+        return PyFloat_FromDouble(sec); \
+    }
+
+STATX_GET_DOUBLE(st_atime, atime_sec, STATX_ATIME)
+STATX_GET_DOUBLE(st_birthtime, btime_sec, STATX_BTIME)
+STATX_GET_DOUBLE(st_ctime, ctime_sec, STATX_CTIME)
+STATX_GET_DOUBLE(st_mtime, mtime_sec, STATX_MTIME)
+
+#define STATX_GET_NSEC(ATTR, MEMBER, MASK) \
+    static PyObject* \
+    pystatx_result_get_##ATTR(PyObject *op, void *context) \
+    { \
+        Py_statx_result *self = Py_statx_result_CAST(op); \
+        if (!(self->stx.stx_mask & MASK)) { \
+            Py_RETURN_NONE; \
+        } \
+        struct statx_timestamp *ts = &self->stx.MEMBER; \
+        _posixstate *state = PyType_GetModuleState(Py_TYPE(op)); \
+        assert(state != NULL); \
+        return stat_nanosecond_timestamp(state, ts->tv_sec, ts->tv_nsec); \
+    }
+
+STATX_GET_NSEC(st_atime_ns, stx_atime, STATX_ATIME)
+STATX_GET_NSEC(st_birthtime_ns, stx_btime, STATX_BTIME)
+STATX_GET_NSEC(st_ctime_ns, stx_ctime, STATX_CTIME)
+STATX_GET_NSEC(st_mtime_ns, stx_mtime, STATX_MTIME)
 
-/* The low 16 bits of the context pointer are the offset from the start of
```

Comment:
I know the mask bit itself won't fit in the context pointer on 32-bit, but you could store the mask bit _index_ in the upper 16 bits.  For example, `STATX_ATIME` is `0x20`, so you'd store `5` in the upper 16 bits, and recover the mask bit as `1 << i` in the getter.  That way you'd only need one getter per type, rather than one per member, which is a lot less code.

Reply:
Oh ok. The problem is that I'm not sure that all compilers supported by Python on Linux can compute the bit position at build time.

I don't think that code size is an issue here.